docs: Integration section and tutorial (#17132)
eslint/eslint: docs/src/integrate/integration-tutorial.md

Author: @nzakas before putting this tutorial up for formal review, do you like the approach taken here? any higher-level feedback you'd like implemented?

* **Custom linter tools**: If you're building a custom linter tool that combines multiple linters or adds specific functionality, you may want to integrate ESLint into your tool to provide JavaScript linting capabilities.

* **Code review tools**: Integrating ESLint with code review tools can help automate the process of identifying potential issues in the codebase.

* **Learning platforms**: If you are developing a learning platform or coding tutorial, integrating ESLint can provide real-time feedback to users as they learn JavaScript, helping them improve their coding skills and learn best practices.

* **Developer tools** If you're creating developer tools, like a testing framework or a bundler, you may want to integrate ESLint to provide linting capabilities.

## What You'll Build

In this guide, you'll create a simple Node.js project that uses the `ESLint` class to lint files and retrieve results.

## Requirements

This tutorial assumes you are familiar with JavaScript and Node.js.

To follow this tutorial, you'll need to have the following:

* Node.js (v12.0.0 or higher)

> * npm or Yarn package manager

Reviewer: Yarn isn't used anywhere on the page, so I think we can safely omit it.
```suggestion
* npm
```

* A text editor

## Step 1: Setup

First, create a new project directory:

```shell
mkdir eslint-integration
cd eslint-integration
```

Initialize the project with a `package.json` file:

```shell
npm init -y
```

> Install the `eslint` package as a dependency:

Reviewer: It may be worth calling out that the package is installed as a dependency and not a dev dependency for this use case.

Author: updated

```shell
npm install eslint
```

Create a new file called `example-eslint-integration.js` in the project root:

```shell

> touch example-eslint-integration.js

Author: no, it wouldn't (well not unless you're using linux subsystem). 

personally i'm comfortable w this as is given the popularity/knowledge of unix-like systems and context of the above text that talks about creating a file. 

but fine adjusting to something like:

```shell
# UNIX-like systems
touch example-eslint-integration.js

# Windows
type NUL > example-eslint-integration.js
```

i feel this is a little clunky, though. and think it's clearer to just keep as is

> touch example-eslint-integration.js

Reviewer: I think `touch` fine, it's fairly well known.

```

## Step 2: Import and Configure the `ESLint` Instance

Import the `ESLint` class from the `eslint` package and create a new instance.

You can customize the ESLint configuration by passing an options object to the `ESLint` constructor:

```javascript
// example-eslint-integration.js

const { ESLint } = require("eslint");

// Create an instance of ESLint with the configuration passed to the function
function createESLintInstance(overrideConfig){
  return new ESLint({ useEslintrc: false, overrideConfig: overrideConfig });
}
```

> ## Step 3: Get Linting Results

Reviewer: We're not just retrieving results here, so maybe this would be a better title?

```suggestion
## Step 3: Lint and Fix Files
```

To lint a file, use the `lintFiles` method of the `ESLint` instance. The `filePaths` argument passed to `ESLint#lintFiles()` can be a string or an array of strings, representing the file path(s) you want to lint. The file paths can be globs or filenames.

> The static method `ESLint.outputFixes()` takes the linting results from the call to `ESLint#lintFiles()`, and then writes any automatic fixes to the source files.

Reviewer: ```suggestion
The static method `ESLint.outputFixes()` takes the linting results from the call to `ESLint#lintFiles()`, and then writes the fixed code back to the source files.
```

```javascript
// example-eslint-integration.js

// ... previous step's code to instantiate the ESLint instance

// Lint the specified files and return the error results
async function lintAndFix(eslint, filePaths) {
  const results = await eslint.lintFiles(filePaths);

  // Apply automatic fixes and output fixed code
  await ESLint.outputFixes(results);

  return results;
}
```

> ## Step 4: Handle Results

Reviewer: ```suggestion
## Step 4: Output Results
```

        env: {
            es6: true,
            node: true,
        },
        parserOptions: {
            ecmaVersion: 2018,
        },
        rules: {
            "no-console": "error",
            "no-unused-vars": "warn",
        },
    };

    const eslint = createESLintInstance(overrideConfig);
    const results = await lintAndFix(eslint, filePaths);
    return outputLintingResults(results);
}

// Export integration
module.exports = { lintFiles }
```

Here's the complete code example for `example-eslint-integration.js`:

```javascript
const { ESLint } = require("eslint");

// Create an instance of ESLint with the configuration passed to the function
function createESLintInstance(overrideConfig){
  return new ESLint({ useEslintrc: false, overrideConfig: overrideConfig });
}

// Lint the specified files and return the error results
async function lintAndFix(eslint, filePaths) {
  const results = await eslint.lintFiles(filePaths);

  // Apply automatic fixes and output fixed code
  await ESLint.outputFixes(results);

  return results;
}

// Log results to console if there are any problems
function outputLintingResults(results) {
  if (results.length) {

>     console.log("Linting errors found!");

Author: in this case, could you please provide an example of what would be correct here?

>     console.log("Linting errors found!");

Author: adjusting. pls lmk what you think of the new approach

    console.log(results);
  } else {
    console.log("No linting errors found.");
  }
  return results;
}

// Put previous functions all together
async function lintFiles(filePaths) {

    // The ESLint configuration. Alternatively, you could load the configuration
    // from a .eslintrc file or just use the default config.
    const overrideConfig = {
        env: {
            es6: true,
            node: true,
        },
        parserOptions: {
            ecmaVersion: 2018,
        },
        rules: {
            "no-console": "error",
            "no-unused-vars": "warn",
        },
    };

    const eslint = createESLintInstance(overrideConfig);
    const results = await lintAndFix(eslint, filePaths);
    return outputLintingResults(results);
}

// Export integration
module.exports = { lintFiles }
```

## Conclusion

In this tutorial, we have covered the essentials of using the `ESLint` class to lint files and retrieve results in your projects. This knowledge can be applied to create custom integrations, such as code editor plugins, to provide real-time feedback on code quality.

## View the Tutorial Code

> You can view the annotated source code for the tutorial [here](https://github.com/eslint/eslint/tree/main/docs/_integration-tutorial-code).

Reviewer: Rather than creating a top-level directory under `docs`, maybe we should have `docs/_examples` for all examples and then a subdirectory under `docs/_examples` for the integration example code?

Author: moved

---
title: Integrate with the Node.js API Tutorial
eleventyNavigation:
    key: integrate with the node.js api tutorial
    parent: integrate eslint
    title: Integrate with the Node.js API Tutorial
    order: 1
---

This guide walks you through integrating the `ESLint` class to lint files and retrieve results, which can be useful for creating integrations with other projects.

> ## Why Create a Custom Integration?

Reviewer: We don't use the word "custom" anywhere else.

```suggestion
## Why Create an Integration?
```

You might want to create an ESLint integration if you're creating developer tooling, such as the following:

- **Code editors and IDEs**: Integrating ESLint with code editors and IDEs can provide real-time feedback on code quality and automatically highlight potential issues as you type. Many editors already have ESLint plugins available, but you may need to create a custom integration if the existing plugins do not meet your specific requirements.

> - **Continuous Integration (CI) pipelines**: Including ESLint in CI pipelines can help enforce code quality standards before merging changes into the main codebase. By integrating ESLint into your CI process, you can automatically run linting checks on pull requests and provide feedback to developers about potential issues.

Reviewer: I'd need to double-check, but I think most CI systems use the CLI.

Author: ok, i can remove this bullet point, then

- **Custom linter tools**: If you're building a custom linter tool that combines multiple linters or adds specific functionality, you may want to integrate ESLint into your tool to provide JavaScript linting capabilities.

- **Code review tools**: Integrating ESLint with code review tools can help automate the process of identifying potential issues in the codebase.

- **Learning platforms**: If you are developing a learning platform or coding tutorial, integrating ESLint can provide real-time feedback to users as they learn JavaScript, helping them improve their coding skills and learn best practices.

> 

Reviewer: I'd also add plugins for code-related tools like Webpack and Jest.

Author: adding a line for 'developer tools'

Reviewer: I think we're still missing the Webpack/Jest plugin use case. In that case you're not creating a new tool but rather extending an existing tool. I think we need to be more explicit about this use case.

Author: updated the bullet point. lmk what you think of the update

First, create a new project directory:

```shell
mkdir eslint-integration
cd eslint-integration
```

Initialize the project with a `package.json` file:

```shell
npm init -y
```

Install the `eslint` package as a dependency:

```shell
npm install eslint
```

Create a new file called `example-eslint-integration.js` in the project root:

```shell
touch example-eslint-integration.js
```

## Step 2: Import and Configure the `ESLint` Instance

Import the `ESLint` class from the `eslint` package and create a new instance. 

You can customize the ESLint configuration by passing an options object to the `ESLint` constructor:

```javascript
// example-eslint-integration.js

const { ESLint } = require("eslint");

function createESLintInstance(eslintConfig){
    if (eslintConfig) {
        return new ESLint({ useEslintrc: false, overrideConfig: eslintConfig });
    } else {
        return new ESLint({ useEslintrc: false });
    }
}
```

## Step 3: Get Linting Results

To lint a file, use the `lintFiles` method of the `ESLint` instance:

The `filePaths` argument passed to `ESLint#lintFiles()` can be a string or an array of strings, representing the file path(s) you want to lint.

> To format the linting results for better readability, use the `outputFixes` and `getErrorResults` methods:

Reviewer: This text feels a bit disjointed to me and disconnected from the following code example. Can you expand this to talk a bit more about what each method does? For instance, lintFiles searches the filesystem and the file paths can be globs or filenames.

Also, outputFixes isn't for readability, it's the method that outputs fixes back to disk. getErrorResults filters out all warnings.

Author: updated copy here

```javascript
// example-eslint-integration.js

// ... previous step's code to instantiate the ESLint instance

/// Define a function that lints the specified files and returns the error results
async function getLintingResults(eslint, filePaths) {
  const results = await eslint.lintFiles(filePaths);

  // Apply automatic fixes and output fixed code
  await ESLint.outputFixes(results);

  // Get error results
  const errorResults = ESLint.getErrorResults(results);

  return errorResults;
}
```

## Step 4: Handle Results

> Define a custom function to handle the linting results. This should be custom to your integration's needs.

Reviewer: Again, "custom" isn't necessary. 
```suggestion
Define a function to output the linting results to the consoles. This should be specific to your integration's needs.
```
